### PR TITLE
Create TermInfoBuf and FdBuf

### PR DESCRIPTION
**Pull request by V-FEXrt**
This should be the final prep PR and the next PR will include actually emitting color in `--failed` and the like.

This PR introduces TermInfoBuf, FdBuf, and the associated tests but does not directly use any of them.

### CODE REVIEW

File: src/runtime/term.cpp [hunk 1]
```diff
@@ -23,9 +23,325 @@
...
+        put(c);
+      }
+      break;
+    case State::esc_state:
+      // TODO: Add other things to this state,
+      // specifically all the escape sequences we
+      // we want to ignore. There should be two
+      // cases, one character cases and two character
+      // cases. For the two character case we'll need
+      // an extra case.
```
Comment (JakeSiFive) [inline src/runtime/term.cpp:275]: I think this todo is todone

File: src/runtime/term.cpp [hunk 2]
```diff
@@ -23,9 +23,325 @@
...
+
+int FdBuf::overflow(int c) {
+  if (c == EOF) {
+    return EOF;
+  }
+
+  char ch = c;
+  ssize_t r = write(fd, &ch, sizeof(ch));
+  if (r == -1) {
+    // TODO(jake): what should be done here?
```
Comment (JakeSiFive) [inline src/runtime/term.cpp:337]: We should just roll over and die. The only error that isn't catastrophic here is `EINTR` so what you could do if you wanted to be smart is this

```
do {
  size_t r = write(fd, &ch, sizeof(ch));
  if (r == -1 && errno != EINTR) { /* roll over and die in a hole */ }
} while (errno != EINTR);
```

That's a maximally safe way to do the write I suppose.

File: src/runtime/term.cpp [hunk 3]
```diff
@@ -58,42 +402,47 @@ const char *term_intensity(int code) {
 
 const char *term_normal() { return sgr0 ? sgr0 : ""; }
 
-bool term_init(bool tty_) {
+bool term_init(bool tty_, bool skip_atty) {
```
Comment (JakeSiFive) [inline src/runtime/term.cpp:405]: please put a todo to reconsider skip_atty after full color support is added.

**Review by JakeSiFive — APPROVED** [review]

File: src/runtime/term.cpp [hunk 4]
```diff
@@ -23,9 +23,323 @@
...
+
+static void write_num(std::streambuf &buf, int x) {
+  std::string s = std::to_string(x);
+  buf.sputn(s.c_str(), s.size());
+}
+
+// In case we have to back up from the num_state to default_state,
+// we put all previous characters. This happens if at some point
+// we have to flush all thus-far collected numbers because the
+// escape code is invalid. There is some special edge case handeling
```
Comment (mwachs5) [inline src/runtime/term.cpp:233]:
```suggestion
// escape code is invalid. There is some special edge case handling
```

**Pull request merged**